### ISSUE TITLE
ES6 class padding

### ISSUE THREAD

**Issue by christianharke** [eslint/eslint#5092]
We would like to enforce padding for ES6 classes. Due to our styleguide those are treated differently than other blocks, which shouldn't be padded:

```js
class MyClass {
    myMethod() {
    }
}
```

and

```js
class MyClass {

    myMethod() {
    }

}
```

Is it intended to add a rule/option for treating them separately?

**Comment by eslintbot** [comment]
@christianharke Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by btmills** [comment]
This convention seems pretty common. Are class bodies sufficiently visually similar to blocks that adding an option to padded-blocks makes sense, or are they semantically different enough that it should be a separate rule? I'm leaning toward the former, but I could see arguments either way. Thoughts?

**Comment by nzakas** [comment]
I'm thinking the former, as well.

**Comment by btmills** [comment]
Sounds good. How about this?

Current options:

```js
module.exports.schema = [
    {
        "enum": ["always", "never"]
    }
];
```

New options:

```js
module.exports.schema = {
    "oneOf": [
        {
            "enum": ["always", "never"]
        },
        {
            "type": "object",
            "properties": {
                "blocks": {
                    "enum": ["always", "never"]
                },
                "classes": {
                    "enum": ["always", "never"]
                }
            },
            "additionalProperties": false,
            "minProperties": 1
        }
    ]
};
```

This retains backward compatibility for the `[2, "always/never"]` form, which would only check blocks and not classes. It also allows blocks and classes to be configured separately: for example, I would configure this as `[2, { "blocks": "never", "classes": "always" }]`. Both `blocks` and `classes` are optional, though at least one must be specified. If either is unset, the corresponding structure is not checked.

**Comment by ilyavolodin** [comment]
Would it be better if backwards compatible option configured both at the same time, and if you wanted finer control, you could do either?

**Comment by nzakas** [comment]
@ilyavolodin that would make it not backwards compatible anymore. :)

**Comment by alberto** [comment]
@nzakas It depends on what do you consider `backwards compatible`. The rule config is still compatible, meaning it won't stop working for people using it. But it will now report more errors now than it did previously. We already do this for many rules, right? Even for this rule, see https://github.com/eslint/eslint/issues/5056

I also think it would be surprising that `always` wouldn't mean always, but just for `blocks`.

**Comment by nzakas** [comment]
Backwards compatible means it works the same way with the exception of bug fixes. If we start flagging things we didn't because we are expanding the scope of the rule, that's no longer backwards compatible.

We can make "always" flag classes too, but that's a breaking change because we are expanding the scope of the rule and changing the meaning of "always". Class bodies are not block statements even though they look like it, so it's hard to rationalize this as a bug fix.

So we can do it in two ways: a breaking way or a non-breaking way.

**Comment by ilyavolodin** [comment]
non-breaking fix sounds good to me, in that case.

**Comment by alberto** [comment]
Well, I'm not sure what our policy is then, if any. Do we strive for correctness (as mentioned above, IMO the expected behavior for `always` in any rule would be to apply to every possible case) or try to avoid breaking changes that would cause users to adjust their rule config?

**Comment by btmills** [comment]
#5162 (fixes #5056) is waiting on a decision here.

@alberto To me, blocks, switches, and classes are separate, albeit visually similar cases. I would want to be able to configure the three separately, so for my uses, `always` suddenly applying to switches and class bodies would be incorrect and a breaking change. That it doesn't check them right now is not a bug but a feature for me. I could see a case for renaming the rule since the other two aren't technically blocks in the ESTree sense, but visually they're blocks of code surrounded by curly braces, so I think it should remain here.

**Comment by nzakas** [comment]
My point is that we can do a breaking change for 2.0.0 if we think that's the right thing to do, I'm just not sure if it is. 

:-1: for renaming the rule. I think we've hit our quota for such changes in 2.0.0.

**Comment by alberto** [comment]
In my view, if it were a new rule we were introducing, `always` should apply to everything. So, that's the "right thing" to do, for me. @btmills that would be a breaking change, yes, but it doesn't mean we don't add the additional options you mentioned for people that want different style for classes.

@eslint/eslint-team thoughts?

**Comment by btmills** [comment]
I think I see the other side here. `"always"` does everything, and if that's not what you want, you can instead configure each individually. That makes sense to me after some thought.

What are the back-compat impacts here? Is making the breaking change even an option at this point?

**Comment by alberto** [comment]
Yeah, I am afraid 2.0 will get released before we get any decision here and this is relegated to 3.0

**Comment by nzakas** [comment]
Or we just do the nonbreaking change for now.

**Comment by btmills** [comment]
So for now, we have this?

```js
module.exports.schema = {
    "oneOf": [
        {
            "enum": ["always", "never"]
        },
        {
            "type": "object",
            "properties": {
                "blocks": {
                    "enum": ["always", "never"]
                },
                "classes": {
                    "enum": ["always", "never"]
                },
                "switches": {
                    "enum": ["always", "never"]
                }
            },
            "additionalProperties": false,
            "minProperties": 1
        }
    ]
};
```

Where `"padded-blocks": [2, "always"]` covers only blocks (not class bodies, not switch statements), and opt-in options to individually configure other types.

Last question: `blocks`/`classes`/`switches` or `block`/`class`/`switch` for the option names?

**Comment by alberto** [comment]
@btmills yes, that should be it. 

Plural seems more consistent with other rules (e.g. `no-use-before-define`)

**Comment by alberto** [comment]
So, is this accepted?

I intended to do this.

**Comment by ljharb** [comment]
+1, would love to see this added

**Comment by nzakas** [comment]
Go for it
